### PR TITLE
[POAE7-2688] String op workflow in nextgen and substring impl.

### PR DESCRIPTION
**Pull request by jikunshang**
### What changes were proposed in this pull request?

Add string op workflow for nextgen compiler, and add substring function impl.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by jikunshang** [comment]
@bigPYJ1151 @YBRua @yma11 Please take a review, thanks.

File: src/cider/type/plan/BinaryExpr.cpp [hunk 1]
```diff
@@ -39,7 +39,9 @@ JITExprValue& BinOper::codegen(JITFunction& func) {
 
   const auto& lhs_ti = lhs->get_type_info();
   const auto& rhs_ti = rhs->get_type_info();
-  CHECK_EQ(lhs_ti.get_type(), rhs_ti.get_type());
+  if (!lhs_ti.is_string()) {
```
Comment (YBRua) [inline src/cider/type/plan/BinaryExpr.cpp:42]: I think, (though it is not very important), for completeness and type safety, when `lhs_ti` is a string type, perhaps we should type-check that `rhs_ti` is also a string type?

File: src/cider/function/scalar/RuntimeFunctions.cpp [hunk 2]
```diff
@@ -1075,6 +1075,7 @@ extern "C" ALWAYS_INLINE uint64_t string_pack(const int8_t* ptr, const int32_t l
 }
 
 #ifdef __clang__
+#include "function/string/CiderStringFunctions.cpp"
```
Comment (yma11) [inline src/cider/function/scalar/RuntimeFunctions.cpp:1078]: Better to put these new functions under nextgen/function folder?

File: src/cider/exec/nextgen/context/RuntimeContext.h [hunk 3]
```diff
@@ -29,12 +29,16 @@
 namespace cider::exec::nextgen::context {
 class RuntimeContext {
  public:
-  RuntimeContext(int64_t ctx_num) : runtime_ctx_pointers_(ctx_num, nullptr) {}
+  RuntimeContext(int64_t ctx_num) : runtime_ctx_pointers_(ctx_num, nullptr) {
+    string_heap_ptr_ = std::make_shared<StringHeap>();
```
Comment (yma11) [inline src/cider/exec/nextgen/context/RuntimeContext.h:33]: Maybe the string heap should initialize during codegen if need? like call `runtime_ctx.setStringHeap()`.

File: src/cider/function/scalar/RuntimeFunctions.cpp (hunk 2 above)
Comment (jikunshang) [reply src/cider/function/scalar/RuntimeFunctions.cpp]: updated.

File: src/cider/exec/nextgen/context/RuntimeContext.h (hunk 3 above)
Comment (jikunshang) [reply src/cider/exec/nextgen/context/RuntimeContext.h]: make sense, updated. Please take a look again.

**Review by bigPYJ1151 — APPROVED** [review]

**Review by yma11 — APPROVED** [review]

**Pull request merged**